File: website/src/components/commentFormatting.ts
// Editor-side model for composing inline comment formatting (issue #318).
//
// A textarea only holds plain text, so we track formatting alongside it as a
// per-character attribute array (one entry per UTF-16 code unit, matching JS
// string indexing and the backend's UTF-16 offset contract). Toolbar actions
// set attributes over the current selection; on submit we compress the array
// into the sorted, non-overlapping spans the API expects.

import type { CommentFormatSpan, TextSize } from '../api/types'

export interface CharAttr {
  bold: boolean
  italic: boolean
  size: TextSize
}

export type CharAttrs = CharAttr[]

const DEFAULT_ATTR: CharAttr = { bold: false, italic: false, size: 'normal' }

function isPlain(a: CharAttr): boolean {
  return !a.bold && !a.italic && a.size === 'normal'
}

/** A fresh attribute array of `len` unstyled characters. */
export function emptyAttrs(len: number): CharAttrs {
  return Array.from({ length: len }, () => ({ ...DEFAULT_ATTR }))
}

/**
 * Reconcile the attribute array across a single textarea edit (typing,
 * pasting, or deleting a contiguous selection) by diffing the common prefix
 * and suffix of the old and new text. Inserted characters get default
 * (unstyled) attributes; untouched characters keep theirs. The result always
 * has one entry per character in `newText`.
 */
export function reconcileAttrs(prev: CharAttrs, oldText: string, newText: string): CharAttrs {
  if (oldText === newText) return prev

  const minLen = Math.min(oldText.length, newText.length)
  let prefix = 0
  while (prefix < minLen && oldText[prefix] === newText[prefix]) prefix++

  let suffix = 0
  while (
    suffix < minLen - prefix &&
    oldText[oldText.length - 1 - suffix] === newText[newText.length - 1 - suffix]
  ) {
    suffix++
  }

  const head = prev.slice(0, prefix)
  const tail = prev.slice(oldText.length - suffix)
  const insertedLen = newText.length - suffix - prefix
  const inserted = emptyAttrs(Math.max(0, insertedLen))
  return [...head, ...inserted, ...tail]
}

/** Apply a partial attribute change to every character in [start, end). */
export function applyStyleToRange(
  attrs: CharAttrs,
  start: number,
  end: number,
  change: Partial<CharAttr>,
): CharAttrs {
  if (start >= end) return attrs
  return attrs.map((a, i) => (i >= start && i < end ? { ...a, ...change } : a))
}

/**
 * Toggle a boolean style over [start, end): if every character in the range
 * already has it on, turn it off; otherwise turn it on. Mirrors how a word
 * processor's Bold/Italic buttons behave on a selection.
 */
export function toggleRange(
  attrs: CharAttrs,
  start: number,
  end: number,
  key: 'bold' | 'italic',
): CharAttrs {
  if (start >= end) return attrs
  let allOn = true
  for (let i = start; i < end; i++) {
    if (!attrs[i][key]) {
      allOn = false
      break
    }
  }
  const change: Partial<CharAttr> = key === 'bold' ? { bold: !allOn } : { italic: !allOn }
  return applyStyleToRange(attrs, start, end, change)
}

/** Compress the attribute array into sorted, non-overlapping formatting spans,
 * dropping unstyled runs. Suitable to send straight to the API. */
export function attrsToSpans(attrs: CharAttrs): CommentFormatSpan[] {
  const spans: CommentFormatSpan[] = []
  let i = 0
  while (i < attrs.length) {
    const a = attrs[i]
    if (isPlain(a)) {
      i++
      continue
    }
    let j = i + 1
    while (
      j < attrs.length &&
      attrs[j].bold === a.bold &&
      attrs[j].italic === a.italic &&
      attrs[j].size === a.size
    ) {
      j++
    }
    spans.push({ start: i, end: j, bold: a.bold, italic: a.italic, size: a.size })
    i = j
  }
  return spans
}

/**
 * Produce the spans for a trimmed comment: comments are `.trim()`-ed before
 * sending, which shifts offsets, so we slice the attribute array to the
 * trimmed span of the raw text before compressing. Returns undefined when
 * there is no formatting, so callers can omit the field entirely.
 */
export function spansForTrimmedComment(
  rawText: string,
  attrs: CharAttrs,
): CommentFormatSpan[] | undefined {
  const leading = rawText.length - rawText.trimStart().length
  const trimmedLength = rawText.trim().length
  const sliced = attrs.slice(leading, leading + trimmedLength)
  const spans = attrsToSpans(sliced)
  return spans.length > 0 ? spans : undefined
}

File: android/PositiveOnlySocial/app/src/main/java/com/example/positiveonlysocial/ui/main/CommentFormatting.kt
package com.example.positiveonlysocial.ui.main

import com.example.positiveonlysocial.data.model.CommentFormatSpan

/**
 * Editor-side per-character formatting for composing a comment (issue #318).
 * Mirrors the web/iOS char-attribute model: the toolbar sets attributes over
 * the current selection, and on submit the array is compressed into the sorted,
 * non-overlapping spans the API expects. One entry per UTF-16 code unit (a
 * Kotlin `Char`), so offsets line up with the backend contract.
 */
data class CharStyle(
    val bold: Boolean = false,
    val italic: Boolean = false,
    val size: String = "normal"
) {
    val isPlain: Boolean get() = !bold && !italic && size == "normal"
}

object CommentFormatting {
    fun emptyStyles(count: Int): List<CharStyle> = List(count) { CharStyle() }

    /**
     * Reconcile the style list across a single text edit by diffing the common
     * prefix and suffix. Inserted characters get default (plain) styles;
     * untouched characters keep theirs. The result always has one entry per
     * character in [newText].
     */
    fun reconcile(prev: List<CharStyle>, oldText: String, newText: String): List<CharStyle> {
        if (oldText == newText) return prev
        val styles = if (prev.size == oldText.length) prev else emptyStyles(oldText.length)
        val minLen = minOf(oldText.length, newText.length)
        var prefix = 0
        while (prefix < minLen && oldText[prefix] == newText[prefix]) prefix++
        var suffix = 0
        while (suffix < (minLen - prefix) &&
            oldText[oldText.length - 1 - suffix] == newText[newText.length - 1 - suffix]
        ) {
            suffix++
        }
        val head = styles.subList(0, prefix)
        val tail = styles.subList(oldText.length - suffix, oldText.length)
        val insertedCount = (newText.length - suffix - prefix).coerceAtLeast(0)
        return head + emptyStyles(insertedCount) + tail
    }

    fun applyToRange(
        styles: List<CharStyle>,
        start: Int,
        end: Int,
        transform: (CharStyle) -> CharStyle
    ): List<CharStyle> {
        if (start >= end) return styles
        return styles.mapIndexed { i, style -> if (i in start until end) transform(style) else style }
    }

    /** Toggle bold over [start, end): off if all already bold, else on. */
    fun toggleBold(styles: List<CharStyle>, start: Int, end: Int): List<CharStyle> {
        if (start >= end) return styles
        val allOn = (start until end).all { it < styles.size && styles[it].bold }
        return applyToRange(styles, start, end) { it.copy(bold = !allOn) }
    }

    /** Toggle italic over [start, end): off if all already italic, else on. */
    fun toggleItalic(styles: List<CharStyle>, start: Int, end: Int): List<CharStyle> {
        if (start >= end) return styles
        val allOn = (start until end).all { it < styles.size && styles[it].italic }
        return applyToRange(styles, start, end) { it.copy(italic = !allOn) }
    }

    fun setSize(styles: List<CharStyle>, start: Int, end: Int, size: String): List<CharStyle> =
        applyToRange(styles, start, end) { it.copy(size = size) }

    /** Compress the style list into sorted, non-overlapping spans, dropping
     * unstyled runs. Returns null when there is no formatting. */
    fun toSpans(styles: List<CharStyle>): List<CommentFormatSpan>? {
        val spans = mutableListOf<CommentFormatSpan>()
        var i = 0
        while (i < styles.size) {
            val style = styles[i]
            if (style.isPlain) {
                i++
                continue
            }
            var j = i + 1
            while (j < styles.size && styles[j] == style) j++
            spans.add(CommentFormatSpan(i, j, style.bold, style.italic, style.size))
            i = j
        }
        return spans.ifEmpty { null }
    }
}
